### ISSUE TITLE
Negative test case issues

### ISSUE THREAD

**Issue by yasker** [longhorn/longhorn#1337]
The issues in this epic spread across v1.0.0 and v1.0.1.

**Comment by yasker** [comment]
Fix for https://github.com/longhorn/longhorn/issues/1260 : (from https://github.com/longhorn/longhorn-manager/pull/558)

1. A new API field named `ready` is introduced. It can indicate if the volume is available for workloads now. If the volume is Faulted or is auto attached, the `ready` is false. The reason why auto attached volumes are not usable is that the auto attachment is triggered by one of the following cases:
    1. The volume is restoring data.
    2. This is a DR volume.
    3. The volume is expanding size.
2. Longhorn will mark the replica as failed as long as the related restore status contains an error.
3. If the pre-restore check fails, Longhorn will manually set errors for all replica status.

**Comment by yasker** [comment]
Regarding https://github.com/longhorn/longhorn/issues/1188 : 

> The PR for issue #1260 cannot completely fix this issue. The 1st test can be passed but the 2nd one will still fail after the fix. Since there are some errors during the incremental restore not being recorded in the restore status. Then the longhorn manager cannot mark the related replica as failed.
> The errors are not recorded in the sync agent server restore status:
> 
> Parameters check failure before restore status initialization.
> Delta file check and cleanup error before starting the incremental restoration.
> File coalesce error in the post inc restore function.
> Replica reload error in the post inc restore function.
> ...
> Besides, there is no related integration test covers those corner cases.
> And the regular restore implementation has the similar issue, too.
> 
> If we want to cover those errors in the restore status, we need to check which error may be fatal for the replica (longhorn manager needs to fail the replica according to the restore status error msg) and which error won't affect the replica (longhorn manager doesn't need to fail the replica and simply retrying restore may be enough.). It means we need to refactor the whole restore, including incremental restore. I don't think we have time to do that in this release.
> 
> Actually I can fix the issue that triggered by the reproduce step, but I don't think that's what we want.

**Comment by yasker** [comment]
Fix for https://github.com/longhorn/longhorn/issues/1253 : (PR: https://github.com/longhorn/longhorn-engine/pull/487 )

>If the engine finds the invalid replicas during starting, it will add the invalid replicas into the backend list but with backend client nil and the backend mode ERR, instead of directly erroring out. Then the engine can still work and the invalid replica instances will be shut down by the manager later.

**Comment by shuo-wu** [comment]
Regarding #1243 :

The basic idea:
> 
> After the discussion with Sheng, we suggest that there is no need to fix this issue as long as the following conditions are satisfied:
> 1. The replica can still r/w data after the lazy unmount.
> 2. The data won't be written to the underlying root filesystem after the lazy unmount.
> 3. The replica can become error immediately when the users try to do snapshot/expansion/...
> 4. The written data after the lazy unmount and before the replica becoming error is correct.

@meldafrawi 's Validations:
> There are 3 cases to validate this issue:
> - Case(1/3): Volume is attached to a node, no process is keeping volume mount directory open
> - Case(2/3): Volume is attached to a node, process e.g bash is keeping volume mount directory open
> - Case(3/3): Volume is used by a Kubernetes workload e.g pod.
> 
> For Cases 2 & 3, Validation PASSED, All data md5sum in the mounted volume head file is consistent with the original data md5sum, and i got the same result as @shuo-wu described.
> 
> For Case1 PARTIAL FAILED: Only data written before lazy unmount existed,

I cannot reproduce it but found another issue for the above case 1. (Additional context: Step 8 is to verify the data is still correct after the lazy unmount happens and the replica becomes failed):
> The issue I encounter here is that the file md5sum calculation in step 8 is pretty slow. One 100Mi file can take more than half hour! There is no error either in dmesg log, or in fsck <the volume head file> result. I don't know what's wrong.

**Comment by shuo-wu** [comment]
Regarding #1330 :

> If all replicas are invalid during the engine starting stage, the engine monitor is not launched since the engine instance is still `starting`. The engine process has no way to tell the manager that all replicas should be mode `ERR` and directly errors out. The longhorn manager considers the replicas as valid but the engine dies unexpectedly, then keeps trying auto reattachment. Finally, the volume gets stuck in the detach-reattach loop.

> Even if all replica can be marked as failed, e.g., manually marking all replicas as failed in the volume controller if engine.Status.CurrentState == "Error" && !engine.Status.Started, the auto-salvage feature will try to reuse all replicas. Then the volume still loops in detach and reattach. This is another issue tracked in #1342.

**Comment by shuo-wu** [comment]
Regarding #1342:

> If issue #1330 is fixed, all invalid replicas will become failed as long as the engine cannot use them to startup. But the auto salvage will be triggered in this case, then the volume still gets stuck in the detach-reattach loop.

**Comment by shuo-wu** [comment]
Fix for #1255 (PR https://github.com/longhorn/longhorn-manager/pull/564):

> As we discussed in the meeting, this issue consists of 3 parts:
> 
> The volume should become Faulted IF the restore fails. And this is fixed by the PR of #1260.
> The CSI plugin cannot use the volume if it is not ready for workloads. Actually Longhorn already prevents the CSI plugin to use the restore/DR volume and the expanding volume. Hence we don't need to worry about this part. But I have updated the go client then add v.ready check in the CSI plugin anyway.
> Whether we need to disable the PV/PVC creation for the restoring volume. The conclusion is NO. Since the users can still create PV and PVC via yaml files. And creating PV/PVC in advance for restoring volumes can shorten the time that the users create the workload with the volume.

**Comment by shuo-wu** [comment]
Regarding #1336:

> There are some weird/wrong behaviors after deleting the engine manager:
> 1. Stopping the errored engine takes a long time (about 1 minute).
> 2. The volume will be reattached automatically.
> 3. The volume will become `Ready` after the automatic reattachment and detachment.
> 
> Ideally, the engine should be `stopped `and the volume should be Faulted after the engine manager deletion. And this shouldn't take a long time.

This is a derived negative case from #1260 

Need to do more investigation along with #1270 and #1328.

**Comment by shuo-wu** [comment]
Regarding #1270, there are 2 parts for this issue.
(The PR for the 1st part: https://github.com/longhorn/longhorn-manager/pull/553)

---
The 1st part:
> If there is a replica deleted during the restoration, a new replica will be rebuild. But this rebuild replica won't continue restoring data after the rebuild complete. Hence it's actually an invalid replica. BUG!

> Disable replica rebuild for the restoring volume

The derived issue/improvement: We can enable the rebuild later by directly starting restoring data for new replicas. (#1279)

---
The 2nd part:
The restore volume doesn't act as expected if the node is down during the restore.

> There are actually 2 sub-cases for the node down:
> 1. The restore volume attached node (the engine node) is down.
> 2. The replica node of the restore volume is down.

(For the 2nd case)
>  the replica on the down node should become failed and the restore can be done.

(For the 1st case) 
> I think Longhorn needs to directly mark the volume as Faulted. Since the restore volume won't be auto reattached again after the node is back. And the engine process doesn't know how to reuse the incomplete snapshot to continue the full restore, either. BTW, Longhorn also needs to mark DR volumes as Faulted for this node down case.

This is a derived negative case from #1260

Need to do more investigation along with #1336 and #1328.

**Comment by shuo-wu** [comment]
Regarding #1328:

> I don't see volume stuck in attaching state anymore.
> 
> Rebooted the node where volume was attached. - Volume stuck in attaching/detaching recursive state.
> Rebooted the node that volume was not attached to - DRV completed the restoring with degraded status.

> Originally, the issue was if a node(volume not attached to this node) where a replica was running rebooted while restore was in progress, it was stuck in attaching state forever.
> 
> Now, the above case doesn't have problem but if the attached node is rebooted volume gets stuck in attaching/detaching state.

This is a derived negative case from #1260

Need to do more investigation along with #1336 and #1270.

**Comment by shuo-wu** [comment]
Regarding #1327:

> - While the restore is in progress, Network of one of the nodes is disconnected (100 seconds)
> - The volume is seen in Faulted state

> Expected behavior
> Volume should be in detached state after restoration. Replica on the node with disconnected network, should fail and be in "Stopped" State.

**Comment by innobead** [comment]
remove stale due to the incorrect exempt label

**Comment by innobead** [comment]
Closing in favor of https://github.com/longhorn/longhorn/issues/5240.
